### PR TITLE
[SPARK-36644][SQL] Push down boolean column filter

### PR DESCRIPTION
**Pull request by kazuyukitanimura**
### What changes were proposed in this pull request?
This PR proposes to improve `DataSourceStrategy` to be able to push down boolean column filters. Currently boolean column filters do not get pushed down and may cause unnecessary IO.

### Why are the changes needed?
The following query does not push down the filter in the current implementation
```
SELECT * FROM t WHERE boolean_field
```
although the following query pushes down the filter as expected.
```
SELECT * FROM t WHERE boolean_field = true
```
This is because the Physical Planner (`DataSourceStrategy`) currently only pushes down limited expression patterns like`EqualTo`.
It is fair for Spark SQL users to expect `boolean_field` performs the same as `boolean_field = true`.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added unit tests
```
build/sbt "core/testOnly *DataSourceStrategySuite   -- -z SPARK-36644"
```

### CODE REVIEW

**Comment by kazuyukitanimura** [comment]
cc @dbtsai @sunchao @viirya @dongjoon-hyun

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala [hunk 1]
```diff
@@ -311,6 +312,21 @@ class DataSourceStrategySuite extends PlanTest with SharedSparkSession {
...
 
+  test("SPARK-36644: Push down boolean column filter") {
+    testTranslateFilter('col.boolean, Some(sources.EqualTo("col", true)))
+
+    val t = "test_table"
+    withTable(t) {
+      import testImplicits._
+      Seq(Some(true), Some(false), None).toDF().write.saveAsTable(t)
+      val df = spark.table(t)
+      df.where("value").queryExecution.executedPlan.collectFirst {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala:323]: We should also verify the result after pushdown.

**Comment by dongjoon-hyun** [comment]
Thank you for pinging me, @kazuyukitanimura .

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala (hunk 1 above)
Comment (kazuyukitanimura) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala]: updated and passed all tests

**Review by sunchao — APPROVED** [review]
LGTM (non-binding). cc @cloud-fan too.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala [hunk 2]
```diff
@@ -311,6 +311,22 @@ class DataSourceStrategySuite extends PlanTest with SharedSparkSession {
...
+    withTable(t) {
+      import testImplicits._
+      Seq(Some(true), Some(false), None).toDF().write.saveAsTable(t)
+      val df = spark.table(t).where("value")
+      df.queryExecution.executedPlan.collectFirst {
+        case f: FileSourceScanExec =>
+          assert(f.metadata("PushedFilters") == "[IsNotNull(value), EqualTo(value,true)]")
+      }
+      checkAnswer(df, Row(true))
+    }
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala:327]: hm, do we need e2e test here?

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala:327]: I think we can just remove

**Comment by HyukjinKwon** [comment]
looks fine to me 2

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala:327]: We should follow the existing code in this suite to add a UT, not an end-to-end test

Comment (kazuyukitanimura) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategySuite.scala]: Thank you! Removed

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Review by viirya — APPROVED** [review]

**Review by dbtsai — APPROVED** [review]

**Comment by dbtsai** [comment]
Merged into master. Thanks @kazuyukitanimura

**Pull request closed without merging**